[aotd] Fix rrelu compilation

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136008

Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

IvanKobzarev committed:
[aotd] Fix rrelu compilation

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/opinfo/core.py:
```diff
@@ -896,6 +896,8 @@ class OpInfo:
 
     is_factory_function: bool = False
 
+    is_randomized_result: bool = False
```
Can you say more about why we need this flag? The main reason I'm surprised is:

(1) most ops that are involve randomness can be made deterministic by setting various pytorch state, e.g. `torch.mamul_seed` ([doc](https://pytorch.org/docs/stable/generated/torch.manual_seed.html)). I think we can... probably figure out how to reset the state between running the ref/test so the output? (also, we have other rng ops that don't seem to need a special flag in the OpInfo tests today)

(2) inductor does generate randomness eagerly from eager mode, but the tests you're updating aren't actually using inductor (they use AOTDispatcher, aka they are capturing a graph of the same underlying aten ops and replaying them)

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev committed:
Update on "[aotd] Fix rrelu compilation"


Issues:
https://github.com/pytorch/pytorch/issues/135083
https://github.com/pytorch/pytorch/issues/120292

rrelu decomposition contains mutation, copy_. Decompositions are executed below Functionalization, as a result AOT produces non-functional graph.

Also that decomposition is registered as python_dispatch kernel for AutogradCUDA. 
Autograd dispatch happens above Functionalization, so registering it for Autograd to handle all backends makes functionalization running after this.

Testing:
```
python test/functorch/test_aotdispatch.py -k test_rrelu 
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
@junlin-habana Thanks for the testing, will start debugging.

IvanKobzarev commented:
@pytorchbot merge

IvanKobzarev commented:
@junlin-habana I rechecked your test:
```
    def test_rrelu_noise_mutation(self):
        def fn(x):
            noise = torch.ones_like(x)
            result = torch._C._nn.rrelu_with_noise(x, noise, 0.2, 0.8, True)
            return result, noise
        
        x = -torch.abs(torch.randn(4, 4, requires_grad=True))

        
        ref_y, ref_noise = fn(x)
        print(f"XXX ref_noise")
        print(ref_noise)

        compiled_fn = torch.compile(fn, backend="inductor", fullgraph=True)
        comp_y, comp_noise = compiled_fn(x)
        print(f"XXX comp_noise")
        print(comp_noise)        
```
And noise mutation is captured:
```
XXX ref_noise
tensor([[0.3616, 0.7460, 0.6859, 0.2228],
        [0.6393, 0.5682, 0.3641, 0.3979],
        [0.3886, 0.3187, 0.6494, 0.5895],
        [0.3719, 0.3988, 0.4285, 0.4109]])
XXX comp_noise
tensor([[0.7303, 0.6850, 0.6517, 0.7393],
        [0.6103, 0.6595, 0.7490, 0.4396],
        [0.2660, 0.3525, 0.4600, 0.4670],
        [0.4980, 0.6719, 0.5962, 0.2782]], grad_fn=<CompiledFunctionBackward>)
```
And aot graph also captures the mutation:
```
  def forward(self, primals_1: "f32[4, 4][4, 1]cpu"):
       # File: /data/users/ivankobzarev/b/pytorch/test/functorch/test_aotdispatch.py:6008 in fn, code: result = torch._C._nn.rrelu_with_noise(x, noise, 0.2, 0.8, True)
      le: "b8[4, 4][4, 1]cpu" = torch.ops.aten.le.Scalar(primals_1, 0)
      uniform: "f32[4, 4][4, 1]cpu" = torch.ops.aten.uniform.default(primals_1, 0.2, 0.8)
      mul: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mul.Tensor(primals_1, uniform)
      where: "f32[4, 4][4, 1]cpu" = torch.ops.aten.where.self(le, mul, primals_1);  mul = primals_1 = None
      full_default_1: "f32[][]cpu" = torch.ops.aten.full.default([], 1.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
      where_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.where.self(le, uniform, full_default_1);  full_default_1 = None
      return (where, where_1, le, uniform)

```

Could you please recheck on the latest master?

junlin-habana commented:
> @junlin-habana I rechecked your test:
> 
> ```
>     def test_rrelu_noise_mutation(self):
>         def fn(x):
>             noise = torch.ones_like(x)
>             result = torch._C._nn.rrelu_with_noise(x, noise, 0.2, 0.8, True)
>             return result, noise
>         
>         x = -torch.abs(torch.randn(4, 4, requires_grad=True))
> 
>         
>         ref_y, ref_noise = fn(x)
>         print(f"XXX ref_noise")
>         print(ref_noise)
> 
>         compiled_fn = torch.compile(fn, backend="inductor", fullgraph=True)
>         comp_y, comp_noise = compiled_fn(x)
>         print(f"XXX comp_noise")
>         print(comp_noise)        
> ```
> 
> And noise mutation is captured:
> 
> ```
> XXX ref_noise
> tensor([[0.3616, 0.7460, 0.6859, 0.2228],
>         [0.6393, 0.5682, 0.3641, 0.3979],
>         [0.3886, 0.3187, 0.6494, 0.5895],
>         [0.3719, 0.3988, 0.4285, 0.4109]])
> XXX comp_noise
> tensor([[0.7303, 0.6850, 0.6517, 0.7393],
>         [0.6103, 0.6595, 0.7490, 0.4396],
>         [0.2660, 0.3525, 0.4600, 0.4670],
>         [0.4980, 0.6719, 0.5962, 0.2782]], grad_fn=<CompiledFunctionBackward>)
> ```
> 
> And aot graph also captures the mutation:
> 
> ```
>   def forward(self, primals_1: "f32[4, 4][4, 1]cpu"):
>        # File: /data/users/ivankobzarev/b/pytorch/test/functorch/test_aotdispatch.py:6008 in fn, code: result = torch._C._nn.rrelu_with_noise(x, noise, 0.2, 0.8, True)
>       le: "b8[4, 4][4, 1]cpu" = torch.ops.aten.le.Scalar(primals_1, 0)
>       uniform: "f32[4, 4][4, 1]cpu" = torch.ops.aten.uniform.default(primals_1, 0.2, 0.8)
>       mul: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mul.Tensor(primals_1, uniform)
>       where: "f32[4, 4][4, 1]cpu" = torch.ops.aten.where.self(le, mul, primals_1);  mul = primals_1 = None
>       full_default_1: "f32[][]cpu" = torch.ops.aten.full.default([], 1.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
>       where_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.where.self(le, uniform, full_default_1);  full_default_1 = None
>       return (where, where_1, le, uniform)
> ```
> 
> Could you please recheck on the latest master?

Ok,thanks for you help, looks good on the latest master.

intellinjun commented:
> @junlin-habana I rechecked your test:
> 
> ```
>     def test_rrelu_noise_mutation(self):
>         def fn(x):
>             noise = torch.ones_like(x)
>             result = torch._C._nn.rrelu_with_noise(x, noise, 0.2, 0.8, True)
>             return result, noise
>         
>         x = -torch.abs(torch.randn(4, 4, requires_grad=True))
> 
>         
>         ref_y, ref_noise = fn(x)
>         print(f"XXX ref_noise")
>         print(ref_noise)
> 
>         compiled_fn = torch.compile(fn, backend="inductor", fullgraph=True)
>         comp_y, comp_noise = compiled_fn(x)
>         print(f"XXX comp_noise")
>         print(comp_noise)        
> ```
> 
> And noise mutation is captured:
> 
> ```
> XXX ref_noise
> tensor([[0.3616, 0.7460, 0.6859, 0.2228],
>         [0.6393, 0.5682, 0.3641, 0.3979],
>         [0.3886, 0.3187, 0.6494, 0.5895],
>         [0.3719, 0.3988, 0.4285, 0.4109]])
> XXX comp_noise
> tensor([[0.7303, 0.6850, 0.6517, 0.7393],
>         [0.6103, 0.6595, 0.7490, 0.4396],
>         [0.2660, 0.3525, 0.4600, 0.4670],
>         [0.4980, 0.6719, 0.5962, 0.2782]], grad_fn=<CompiledFunctionBackward>)
> ```
> 
> And aot graph also captures the mutation:
> 
> ```
>   def forward(self, primals_1: "f32[4, 4][4, 1]cpu"):
>        # File: /data/users/ivankobzarev/b/pytorch/test/functorch/test_aotdispatch.py:6008 in fn, code: result = torch._C._nn.rrelu_with_noise(x, noise, 0.2, 0.8, True)
>       le: "b8[4, 4][4, 1]cpu" = torch.ops.aten.le.Scalar(primals_1, 0)
>       uniform: "f32[4, 4][4, 1]cpu" = torch.ops.aten.uniform.default(primals_1, 0.2, 0.8)
>       mul: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mul.Tensor(primals_1, uniform)
>       where: "f32[4, 4][4, 1]cpu" = torch.ops.aten.where.self(le, mul, primals_1);  mul = primals_1 = None
>       full_default_1: "f32[][]cpu" = torch.ops.aten.full.default([], 1.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
>       where_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.where.self(le, uniform, full_default_1);  full_default_1 = None
>       return (where, where_1, le, uniform)
> ```
> 
> Could you please recheck on the latest master?

@IvanKobzarev 
Can you change the dtype of the input x to bfloat16? Like the following, in which case I see that comp_noise is not changed
```
x = -torch.abs(torch.randn(4, 4, dtype=torch.bfloat16, requires_grad=True))
```

IvanKobzarev commented:
@intellinjun Thanks, reproduced. With bf16 it is not mutating noise, smth with dtype convertion in compilation. Will debug it.

IvanKobzarev commented:
@intellinjun https://github.com/pytorch/pytorch/issues/136784 for tracking as a separate issue
